core/consensus/babe/src/lib.rs

```diff
@@ -820,285 +843,298 @@ fn initialize_authorities_cache<B, C>(client: &C) -> Result<(), ConsensusError>
+		SharedEpochChanges {
+			inner: Arc::new(Mutex::new(epoch_changes))
+		}
 	}
+}
 
-	impl TestNetFactory for BabeTestNet {
-		type Specialization = DummySpecialization;
-		type Verifier = BabeVerifier<PeersFullClient>;
-		type PeerData = ();
+/// A block-import handler for BABE.
+///
+/// This scans each imported block for epoch change signals. The signals are
+/// tracked in a tree (of all forks), and the import logic validates all epoch
+/// change transitions, i.e. whether a given epoch change is expected or whether
+/// it is missing.
+///
+/// The epoch change tree should be pruned as blocks are finalized.
+pub struct BabeBlockImport<B, E, Block: BlockT, I, RA> {
+	inner: I,
+	client: Arc<Client<B, E, Block, RA>>,
+	epoch_changes: SharedEpochChanges<Block>,
+}
 
-		/// Create new test network with peers and given config.
-		fn from_config(_config: &ProtocolConfig) -> Self {
-			debug!(target: "babe", "Creating test network from config");
-			BabeTestNet {
-				peers: Vec::new(),
-			}
+impl<B, E, Block: BlockT, I: Clone, RA> Clone for BabeBlockImport<B, E, Block, I, RA> {
+	fn clone(&self) -> Self {
+		BabeBlockImport {
+			inner: self.inner.clone(),
+			client: self.client.clone(),
+			epoch_changes: self.epoch_changes.clone(),
 		}
+	}
+}
 
-		fn make_verifier(&self, client: PeersClient, _cfg: &ProtocolConfig)
-			-> Arc<Self::Verifier>
-		{
-			let client = client.as_full().expect("only full clients are used in test");
-			trace!(target: "babe", "Creating a verifier");
-			let config = Config::get_or_compute(&*client)
-				.expect("slot duration available");
-			let inherent_data_providers = InherentDataProviders::new();
-			register_babe_inherent_data_provider(
-				&inherent_data_providers,
-				config.get()
-			).expect("Registers babe inherent data provider");
-			trace!(target: "babe", "Provider registered");
-
-			assert_eq!(config.get(), SLOT_DURATION);
-			Arc::new(BabeVerifier {
-				client,
-				inherent_data_providers,
-				config,
-				time_source: Default::default(),
-			})
+impl<B, E, Block: BlockT, I, RA> BabeBlockImport<B, E, Block, I, RA> {
+	fn new(
+		client: Arc<Client<B, E, Block, RA>>,
+		epoch_changes: SharedEpochChanges<Block>,
+		block_import: I,
+	) -> Self {
+		BabeBlockImport {
+			client,
+			inner: block_import,
+			epoch_changes,
 		}
+	}
+}
 
-		fn peer(&mut self, i: usize) -> &mut Peer<Self::PeerData, DummySpecialization> {
-			trace!(target: "babe", "Retreiving a peer");
-			&mut self.peers[i]
+impl<B, E, Block, I, RA> BlockImport<Block> for BabeBlockImport<B, E, Block, I, RA> where
+	Block: BlockT<Hash=H256>,
+	I: BlockImport<Block> + Send + Sync,
+	I::Error: Into<ConsensusError>,
+	B: Backend<Block, Blake2Hasher> + 'static,
+	E: CallExecutor<Block, Blake2Hasher> + 'static + Clone + Send + Sync,
+	RA: Send + Sync,
+{
+	type Error = ConsensusError;
+
+	fn import_block(
+		&mut self,
+		mut block: BlockImportParams<Block>,
+		new_cache: HashMap<well_known_cache_keys::Id, Vec<u8>>,
```

Comment:
what is this populated with when the blocks don't come from the authorship task?

Reply:
https://github.com/paritytech/substrate/pull/3028/files#diff-671e8a8d7d8649656f0988fafca31301R689 same thing. the import queue then just create an hashmap and adds those keys in.